File: library/src/com/google/maps/android/geojson/GeoJsonPointStyle.java
package com.google.maps.android.geojson;

import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.Observable;

/**
 * A class that allows for GeoJsonPoint objects to be styled and for these styles to be translated
 * into a MarkerOptions object
 */
public class GeoJsonPointStyle extends Observable implements GeoJsonStyle {

    private final static String GEOMETRY_TYPE_REGEX = "Point|MultiPoint|GeometryCollection";

    private final MarkerOptions mMarkerOptions;


    /**
     * Creates a new PointStyle object
     */
    public GeoJsonPointStyle() {
        mMarkerOptions = new MarkerOptions();
    }

    /**
     * Gets the type of geometries this style can be applied to
     *
     * @return type of geometries this style can be applied to
     */
    @Override
    public String getGeometryType() {
        return GEOMETRY_TYPE_REGEX;
    }

    /**
     * Gets the alpha of the Point
     *
     * @return alpha of the Point
     */
    public float getAlpha() {
        return mMarkerOptions.getAlpha();
    }

    /**
     * Sets the alpha of the GeoJsonPoint
     *
     * @param alpha alpha value of the GeoJsonPoint
     */
    public void setAlpha(float alpha) {
        mMarkerOptions.alpha(alpha);
        styleChanged();
    }

    /**
     * Gets the Anchor U coordinate of the GeoJsonPoint
     *
     * @return Anchor U coordinate of the GeoJsonPoint
     */
    public float getAnchorU() {
        return mMarkerOptions.getAnchorU();
    }

    /**
     * Gets the Anchor V coordinate of the GeoJsonPoint
     *
     * @return Anchor V coordinate of the GeoJsonPoint
     */
    public float getAnchorV() {
        return mMarkerOptions.getAnchorV();
    }

    /**
     * Sets the Anchor U and V coordinates of the GeoJsonPoint
     *
     * @param anchorU Anchor U coordinate of the GeoJsonPoint
     * @param anchorV Anchor V coordinate of the GeoJsonPoint
     */
    public void setAnchor(float anchorU, float anchorV) {
        mMarkerOptions.anchor(anchorU, anchorV);
        styleChanged();
    }

    /**
     * Gets whether the GeoJsonPoint is draggable
     *
     * @return true if GeoJsonPoint is draggable, false if not draggable
     */
    public boolean isDraggable() {
        return mMarkerOptions.isDraggable();
    }

    /**
     * Sets the GeoJsonPoint to be draggable
     *
     * @param draggable true if GeoJsonPoint is draggable, false if not draggable
     */
    public void setDraggable(boolean draggable) {
        mMarkerOptions.draggable(draggable);
        styleChanged();
    }

    /**
     * Gets whether the GeoJsonPoint is flat
     *
     * @return true if GeoJsonPoint is flat, false if not flat
     */
    public boolean isFlat() {
        return mMarkerOptions.isFlat();
    }

    /**
     * Sets the GeoJsonPoint to be flat
     *
     * @param flat true if GeoJsonPoint is flat, false if not flat
     */
    public void setFlat(boolean flat) {
        mMarkerOptions.flat(flat);
        styleChanged();
    }

    /**
     * Gets a bitmap image for the GeoJsonPoint
     *
     * @return bitmap descriptor for the GeoJsonPoint
     */
    public BitmapDescriptor getIcon() {
        return mMarkerOptions.getIcon();
    }

    /**
     * Sets a bitmap image for the GeoJsonPoint
     *
     * @param bitmap bitmap descriptor for the GeoJsonPoint
     */
    public void setIcon(BitmapDescriptor bitmap) {
        mMarkerOptions.icon(bitmap);
        styleChanged();
    }

    /**
     * Gets the info window anchor U coordinate of the GeoJsonPoint
     *
     * @return info window anchor U coordinate of the GeoJsonPoint
     */
    public float getInfoWindowAnchorU() {
        return mMarkerOptions.getInfoWindowAnchorU();
    }

    /**
     * Gets the info window anchor V coordinate of the GeoJsonPoint
     *
     * @return info window anchor V coordinate of the GeoJsonPoint
     */
    public float getInfoWindowAnchorV() {
        return mMarkerOptions.getInfoWindowAnchorV();
    }

    /**
     * Sets the info window anchor U and V coordinates of the GeoJsonPoint
     *
     * @param infoWindowAnchorU info window anchor U coordinate of the GeoJsonPoint
     * @param infoWindowAnchorV info window anchor V coordinate of the GeoJsonPoint
     */
    public void setInfoWindowAnchor(float infoWindowAnchorU, float infoWindowAnchorV) {
        mMarkerOptions.infoWindowAnchor(infoWindowAnchorU, infoWindowAnchorV);
        styleChanged();
    }

    /**
     * Gets the rotation of the GeoJsonPoint
     *
     * @return rotation of the GeoJsonPoint
     */
    public float getRotation() {
        return mMarkerOptions.getRotation();
    }


    /**
     * Sets the rotation of the GeoJsonPoint
     *
     * @param rotation rotation value of the GeoJsonPoint
     */
    public void setRotation(float rotation) {
        mMarkerOptions.rotation(rotation);
        styleChanged();
    }

    /**
     * Gets the snippet of the GeoJsonPoint
     *
     * @return snippet of the GeoJsonPoint
     */
    public String getSnippet() {
        return mMarkerOptions.getSnippet();
    }

    /**
     * Sets the snippet of the GeoJsonPoint
     *
     * @param snippet sets the snippet value of the GeoJsonPoint
     */
    public void setSnippet(String snippet) {
        mMarkerOptions.snippet(snippet);
        styleChanged();
    }

    /**
     * Gets the title of the GeoJsonPoint
     *
     * @return title of the GeoJsonPoint
     */
    public String getTitle() {
        return mMarkerOptions.getTitle();
    }

    /**
     * Sets the title of the GeoJsonPoint
     *
     * @param title title value of the GeoJsonPoint
     */
    public void setTitle(String title) {
        mMarkerOptions.title(title);
        styleChanged();
    }

    /**
     * Gets whether the GeoJsonPoint is visible
     *
     * @return true if GeoJsonPoint is visible, false if not visible
     */
    @Override
    public boolean isVisible() {
        return mMarkerOptions.isVisible();
    }

    /**
     * Sets whether the GeoJsonPoint is visible
     *
     * @param visible true if GeoJsonPoint is visible, false if not visible
     */
    @Override
    public void setVisible(boolean visible) {
        mMarkerOptions.visible(visible);
        styleChanged();
    }

    /**
     * Notifies the observers, GeoJsonFeature objects, that the style has changed. Indicates to the
     * GeoJsonFeature that it should check whether a redraw is needed for the feature.
     */
    private void styleChanged() {
        setChanged();
        notifyObservers();
    }

    /**
     * Gets a new MarkerOptions object containing styles for the GeoJsonPoint
     *
     * @return new MarkerOptions object
     */
    public MarkerOptions toMarkerOptions() {
        MarkerOptions markerOptions = new MarkerOptions();
        markerOptions.alpha(mMarkerOptions.getAlpha());
        markerOptions.anchor(mMarkerOptions.getAnchorU(), mMarkerOptions.getAnchorV());
        markerOptions.draggable(mMarkerOptions.isDraggable());
        markerOptions.flat(mMarkerOptions.isFlat());
        markerOptions.icon(mMarkerOptions.getIcon());
        markerOptions.infoWindowAnchor(mMarkerOptions.getInfoWindowAnchorU(),
                mMarkerOptions.getInfoWindowAnchorV());
        markerOptions.rotation(mMarkerOptions.getRotation());
        markerOptions.snippet(mMarkerOptions.getSnippet());
        markerOptions.title(mMarkerOptions.getTitle());
        markerOptions.visible(mMarkerOptions.isVisible());
        return markerOptions;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("PointStyle{");
        sb.append("\n geometry type=").append(GEOMETRY_TYPE_REGEX);
        sb.append(",\n alpha=").append(getAlpha());
        sb.append(",\n anchor U=").append(getAnchorU());
        sb.append(",\n anchor V=").append(getAnchorV());
        sb.append(",\n draggable=").append(isDraggable());
        sb.append(",\n flat=").append(isFlat());
        sb.append(",\n info window anchor U=").append(getInfoWindowAnchorU());
        sb.append(",\n info window anchor V=").append(getInfoWindowAnchorV());
        sb.append(",\n rotation=").append(getRotation());
        sb.append(",\n snippet=").append(getSnippet());
        sb.append(",\n title=").append(getTitle());
        sb.append(",\n visible=").append(isVisible());
        sb.append("\n}\n");
        return sb.toString();
    }
}

File: data/src/main/java/com/google/maps/android/data/geojson/GeoJsonPointStyle.kt
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.maps.android.data.geojson

import com.google.android.gms.maps.model.BitmapDescriptor
import com.google.android.gms.maps.model.MarkerOptions
import com.google.maps.android.data.Style

/**
 * A class that allows for GeoJsonPoint objects to be styled and for these styles to be translated
 * into a MarkerOptions object.
 */
@Deprecated("Use the new platform-agnostic data layer and renderer instead.")
public class GeoJsonPointStyle :
    Style(),
    GeoJsonStyle {
    override fun getGeometryType(): Array<String> = GEOMETRY_TYPE

    public fun getAlpha(): Float = mMarkerOptions.alpha

    public fun setAlpha(alpha: Float) {
        mMarkerOptions.alpha(alpha)
        styleChanged()
    }

    public fun getAnchorU(): Float = mMarkerOptions.anchorU

    public fun getAnchorV(): Float = mMarkerOptions.anchorV

    public fun setAnchor(
        anchorU: Float,
        anchorV: Float,
    ) {
        setMarkerHotSpot(anchorU, anchorV, "fraction", "fraction")
        styleChanged()
    }

    public fun isDraggable(): Boolean = mMarkerOptions.isDraggable

    public fun setDraggable(draggable: Boolean) {
        mMarkerOptions.draggable(draggable)
        styleChanged()
    }

    public fun isFlat(): Boolean = mMarkerOptions.isFlat

    public fun setFlat(flat: Boolean) {
        mMarkerOptions.flat(flat)
        styleChanged()
    }

    public fun getIcon(): BitmapDescriptor? = mMarkerOptions.icon

    public fun setIcon(bitmap: BitmapDescriptor?) {
        mMarkerOptions.icon(bitmap)
        styleChanged()
    }

    public fun getInfoWindowAnchorU(): Float = mMarkerOptions.infoWindowAnchorU

    public fun getInfoWindowAnchorV(): Float = mMarkerOptions.infoWindowAnchorV

    public fun setInfoWindowAnchor(
        infoWindowAnchorU: Float,
        infoWindowAnchorV: Float,
    ) {
        mMarkerOptions.infoWindowAnchor(infoWindowAnchorU, infoWindowAnchorV)
        styleChanged()
    }

    override fun getRotation(): Float = mMarkerOptions.rotation

    public fun setRotation(rotation: Float) {
        setMarkerRotation(rotation)
        styleChanged()
    }

    public fun getSnippet(): String? = mMarkerOptions.snippet

    public fun setSnippet(snippet: String?) {
        mMarkerOptions.snippet(snippet)
        styleChanged()
    }

    public fun getTitle(): String? = mMarkerOptions.title

    public fun setTitle(title: String?) {
        mMarkerOptions.title(title)
        styleChanged()
    }

    override fun isVisible(): Boolean = mMarkerOptions.isVisible

    override fun setVisible(visible: Boolean) {
        mMarkerOptions.visible(visible)
        styleChanged()
    }

    public fun getZIndex(): Float = mMarkerOptions.zIndex

    public fun setZIndex(zIndex: Float) {
        mMarkerOptions.zIndex(zIndex)
        styleChanged()
    }

    private fun styleChanged() {
        setChanged()
        notifyObservers()
    }

    public fun toMarkerOptions(): MarkerOptions =
        MarkerOptions().apply {
            alpha(mMarkerOptions.alpha)
            anchor(mMarkerOptions.anchorU, mMarkerOptions.anchorV)
            draggable(mMarkerOptions.isDraggable)
            flat(mMarkerOptions.isFlat)
            icon(mMarkerOptions.icon)
            infoWindowAnchor(mMarkerOptions.infoWindowAnchorU, mMarkerOptions.infoWindowAnchorV)
            rotation(mMarkerOptions.rotation)
            snippet(mMarkerOptions.snippet)
            title(mMarkerOptions.title)
            visible(mMarkerOptions.isVisible)
            zIndex(mMarkerOptions.zIndex)
        }

    override fun toString(): String =
        StringBuilder("PointStyle{")
            .apply {
                append("\n geometry type=").append(GEOMETRY_TYPE.contentToString())
                append(",\n alpha=").append(getAlpha())
                append(",\n anchor U=").append(getAnchorU())
                append(",\n anchor V=").append(getAnchorV())
                append(",\n draggable=").append(isDraggable())
                append(",\n flat=").append(isFlat())
                append(",\n info window anchor U=").append(getInfoWindowAnchorU())
                append(",\n info window anchor V=").append(getInfoWindowAnchorV())
                append(",\n rotation=").append(getRotation())
                append(",\n snippet=").append(getSnippet())
                append(",\n title=").append(getTitle())
                append(",\n visible=").append(isVisible())
                append(",\n z index=").append(getZIndex())
                append("\n}\n")
            }.toString()

    companion object {
        private val GEOMETRY_TYPE = arrayOf("Point", "MultiPoint", "GeometryCollection")
    }
}
